feat(autoconfiguration): Annotation based binding

alfonso-presa opened the pull request:
This feature is explained in #4004.

Adds two annotations:
- Configuration: annotates a class provided by modules that returns the
  default bindings for that module
- Autoconfigured: annotates a component so that it gets the bindings
  provided by the imported modules automatically (no need to bind'em during the bootstrap).

For an example of the improvements you can check how the [boostraping is now made in the http example](https://github.com/angular/angular/compare/angular:master...alfonso-presa:autoconfigurationFeature?expand=1#diff-359571929599984fdca28ed5e4d17061R7)

I would really appreciate anyone's feedback about this idea.

I hope you find it useful :-).

alfonso-presa committed:
feat(autoconfiguration): Annotation based binding

Add two annotations:
- Configuration: annotates a class created by modules that returns the
  default bindings for that module
- Autoconfigured: annotates a component so that it gets the bindings
  provided by the imported modules automatically.

Part of: #4004

ericmartinezr commented:
This looks great, and I think for big apps it will be really useful.

I have a question just like an outsider : With the current binding mechanism you can separate the binding depending on what you need for each component. I mean, if CmpA requires only HTTP_BINDINGS, I just pass it to the component through viewBindings/bindings, and any other component won't be aware of that. For what I see in your proposal with `@Autoconfigured()` the component using this annotation will be aware of all `@Configuration()` annotated classes.

I think it will be more easy to understand with an example (using your own examples)

```javascript
// From your example
@Configuration()
class HttpConfiguration {
    getBindings() { return [HTTP_BINDINGS]; }
}

// Skipped everything else from 
// https://github.com/alfonso-presa/angular/blob/465745cf07365dc50c914a8fe111fbbb9bb48cbd/modules/examples/src/hello_world/greeting_service.ts
@Configuration()
class GreetingServiceConfiguration {
  getBindings() { return GREETINGS_BINDINGS; }
}

@Autoconfigured() 
class MyBootstrappableCmp {
 
}
bootstrap(MyBootstrappableCmp);
```

If I understood correctly `MyBootstrappableCmp` would be aware of both bindings GREETINGS and HTTP, but I don't want my component to be aware of every configuration, just what it needs. Is there anyway to separate the configuration? Say something like : "I'll use `@Autoconfigured()` but it will take just these specifics `@Configurations()` and this other `@Autoconfigured()` will use another specifics `@Configurations()`".

Second and last question : is there any difference between returning the bindings within `[]` (http example) or not (greeting example)?

I know this is not up to me, but LGTM :+1:

alfonso-presa commented:
@ericmartinezr Thank you for the quick feedback.

You're right about your first question but there's a little detail to highlight. The HTTP configuration will only be loaded if you actually load the http modules in the application. But it's true that once is loaded, all the components that are *autoconfigured* will get that binding automatically. To solve this I would implement in the future (in case this PR get's accepted of course) some kind of profile system for the components and the configurations, so you can for example get `JSONP_BINDINGS` instead of `HTTP_BINDINGS` depending on your component profile.

BTW I've implemented the `HTTP_BINDINGS` configuration just as showcase, but I think this will be also very useful for the router.

About the second question, there's no difference... In fact I think I should remove those braces from the `HTTP_BINDING`.

Thanks!

mhevery commented:
Hi @alfonso-presa you did put a lot of effort to this PR. I need to chat with @vsavkin about this. My concern is that this kind of magic will break minification, and tree shaking tools, and it most certainly will break transformers as they are now. Let's get a conversation going about the problem, before we jump into a solution.

alfonso-presa commented:
Thank you for your response @mhevery. Please don't hesitate closing this PR if I did something horrible :-P. I did it mainly because I wanted to learn (and also to help improve Angular 2 a little bit if possible), and to play with this *master*piece of software is always fun.

About your concerns (which I totally understand given I modified/added 20 files in a row...),  I don't see how this PR may affect minification or tree shaking as it's still necessary to import the *autoconfigured* component and use some of it's interfaces, otherwise the `Autoconfigured` annotation will not be loaded and then, it will never be discovered, so traceability is still warranted even without the explicit bindings. I modified two e2e tests (`hello-world` and `http`) and both are working nicely in JS and the first in Dart (the second is not transformed) without too much tweaking.

Also, just to clarify, If you're talking about the *autobootstrapping* that I talk about in #4004, then it's true that it will not be possible to implement such a feature without changing the transformers, but *autobootstraping* is not included in this PR (I had it implemented but I dropped it because as said the transformers kept rejecting to include the files if they where not used... they are clever, aren't they? :-) ).

Any way, I agree this is a deep change that might need more discussion before getting accepted (or rejected of course).

Thank you!!

tbosch commented:
This is an interesting feature, thanks. We are currently in the process of not making breaking changes. We can visit it later though...
